lib/mongo/socket/ssl.rb

```diff
@@ -184,6 +187,89 @@ def readbyte
+      def connect_without_timeout(sockaddr)
+        @tcp_socket.connect(sockaddr)
+        @socket = OpenSSL::SSL::SSLSocket.new(@tcp_socket, context)
+        @socket.hostname = @host_name
+        @socket.sync_close = true
+        @socket.connect
+        verify_certificate!(@socket)
+        verify_ocsp_endpoint!(@socket)
+      end
+
+      # Connects the socket with the connect timeout. The timeout applies to
+      # connecting both ssl socket and the underlying tcp socket.
+      #
+      # @param [ String ] sockaddr Address to connect to.
+      def connect_with_timeout(sockaddr, connect_timeout)
+        if connect_timeout <= 0
+          raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+        end
+
+        deadline = Utils.monotonic_time + connect_timeout
+        # Connect tcp socket first.
+        begin
+          @tcp_socket.connect_nonblock(sockaddr)
+        rescue IO::WaitWritable
+          select_timeout = deadline - Utils.monotonic_time
+          if select_timeout <= 0
+            raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+          end
+          if IO.select(nil, [@tcp_socket], nil, select_timeout)
+            retry
+          else
+            raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+          end
+        rescue Errno::EISCONN
+          # Socket is connected, nothing to do.
+        end
+
+        if deadline <= Utils.monotonic_time
+          raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+        end
+
+        @socket = OpenSSL::SSL::SSLSocket.new(@tcp_socket, context)
+        @socket.hostname = @host_name
+        @socket.sync_close = true
+
+        # We still have time, connecting ssl socket.
+        begin
+          @socket.connect_nonblock
+        rescue IO::WaitReadable, OpenSSL::SSL::SSLErrorWaitReadable
+          select_timeout = deadline - Utils.monotonic_time
+          if select_timeout <= 0
+            raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+          end
+          rv = IO.select([@socket], nil, nil, select_timeout)
+          if rv.nil?
+            raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+          else
+            retry
+          end
+        rescue IO::WaitWritable, OpenSSL::SSL::SSLErrorWaitWritable
+          select_timeout = deadline - Utils.monotonic_time
+          if select_timeout <= 0
+            raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+          end
+          rv = IO.select(nil, [@socket], nil, select_timeout)
+          if rv.nil?
+            raise Error::SocketTimeoutError, "The socket took over #{connect_timeout} seconds to connect"
+          else
+            retry
+          end
+          retry
```

Comment:
Looks like a stray `retry` here. Also, feels like these two `rescue` blocks have a lot in common (differ only on whether the sockets are waiting for read or write). Could the logic be pulled out somewhere?

Reply:
Good catch, thank you. Extracted the retry logic in a separate method.